### PR TITLE
job controller: set job conditions to false explicitly

### PR DESCRIPTION
**Pull request by adtac**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/pull/98727#discussion_r588642977

From the guide to writing controllers in the [API conventions doc]:

>Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is Unknown. This allows other components in the system to know that the condition exists ...

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

This is a user-facing change. If external tools were incorrectly relying on the presence/absence of a particular condition instead of its status, such tools will need to correct their behaviour. Whether or not this is acceptable, I'll leave it to the API reviewers.

However, I'll note that other controllers also make this mistake of removing conditions instead of setting the status to false. Non-exhaustive list of examples:

https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/controller/deployment/progress.go#L42

https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/controller/replicaset/replica_set_utils.go#L120

#### Does this PR introduce a user-facing change?

```release-note
The Job controller now sets Job conditions with a false status value when it previously omitted them
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

cc @alculquicondor @lavalamp @soltysh 
/sig apps

### CODE REVIEW

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -986,13 +996,21 @@ func ensureJobConditionStatus(list []batch.JobCondition, cType batch.JobConditio
 				list[i].Reason = reason
```
[the remark this answers is not in the document]
Comment (adtac) [reply pkg/controller/job/job_controller.go:996]: So I did begin to change LastProbeTime, but there's a problem with that. By definition, LastProbeTime must be updated every time ensureJobConditionStatus is called regardless of whether the status/reason/message matches (because it's just a probe). That means that jobConditionsChanged is always true every time syncJob is called. That means that we'll do a `Jobs.UpdatedStatus()` every time syncJob is called and that `forget` is always true. These are significantly more far-reaching consequences than I had originally anticipated.

The performance impact of syncing the Job with the API server every time syncJob is called could be disastrous for long running jobs that see very infrequent updates.

So I'd like to hold off from making LastProbeTime change in this PR. It also sends the wrong message to update it only when job condition status changes. AFAICT there are also no other k8s controllers that do anything like LastProbeTime, so I think removing that field might even be the best course of action.

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -579,38 +582,48 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+			jobConditionsChanged = jobConditionsChanged || isUpdated
 			now := metav1.Now()
 			job.Status.CompletionTime = &now
 			jm.recorder.Event(&job, v1.EventTypeNormal, "Completed", "Job completed")
-		} else if utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && manageJobCalled {
+		} else {
+			job.Status.Conditions, isUpdated = ensureJobConditionStatus(job.Status.Conditions, batch.JobComplete, v1.ConditionFalse, "", "")
+			jobConditionsChanged = jobConditionsChanged || isUpdated
+		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && manageJobCalled {
```
[the remark this answers is not in the document]
Comment (adtac) [reply pkg/controller/job/job_controller.go:594]: Done.

File: pkg/controller/job/job_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:997]: I was thinking that if the block in line 643 is reached, you can update all `LastProbeTime`s

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -579,39 +582,50 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
-					// done because the ActiveDeadlineSeconds timer shouldn't go off
-					// whilst the Job is still suspended and resetting StartTime is
-					// consistent with resuming a Job created in the suspended state.
-					// (ActiveDeadlineSeconds is interpreted as the number of seconds a
-					// Job is continuously active.)
-					now := metav1.Now()
-					job.Status.StartTime = &now
-				}
+		} else {
+			job.Status.Conditions, isUpdated = ensureJobConditionStatus(job.Status.Conditions, batch.JobComplete, v1.ConditionFalse, "", "")
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:591]: why is this necessary?
Wouldn't it be overridden below?

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -579,39 +582,50 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+			jobConditionsChanged = jobConditionsChanged || isUpdated
+		}
+
+		if !utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:595]: Consider putting all this logic in a function.

For example, you could just return after removing the condition if the gate is disabled, simplifying the conditionals below.

File: pkg/controller/job/job_controller.go [hunk 5]
```diff
@@ -579,39 +582,50 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+			job.Status.Conditions, isUpdated = removeJobCondition(job.Status.Conditions, batch.JobSuspended)
+			jobConditionsChanged = jobConditionsChanged || isUpdated
+		} else if !hasCondition(job.Status.Conditions, batch.JobSuspended) {
+			// The feature gate is enabled, but the condition doesn't exist. This
+			// should never happen, insert the default placeholder with status false.
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:600]: wouldn't this happen in the first sync?

File: pkg/controller/job/job_controller_test.go [hunk 6]
```diff
@@ -654,9 +643,14 @@ func TestControllerSyncJob(t *testing.T) {
 			} else if tc.fakeExpectationAtCreation > 0 {
 				manager.expectations.ExpectCreations(key, int(tc.fakeExpectationAtCreation))
 			}
+			originalConditions := []batch.JobCondition{
+				newCondition(batch.JobComplete, v1.ConditionFalse, "", ""),
+				newCondition(batch.JobFailed, v1.ConditionFalse, "", ""),
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:648]: also `newCondition(batch.JobSuspended, v1.ConditionFalse, "", "")`, if feature is enabled.

But, I would rather prefer we have `wantConditions []JobCondition` in the test case.

File: pkg/controller/job/job_controller_test.go [hunk 7]
```diff
@@ -747,16 +741,14 @@ func TestControllerSyncJob(t *testing.T) {
...
-				if cond := hasTrueCondition(actual); cond != nil {
-					t.Errorf("Got condition %s, want none", *cond)
-				}
+				originalConditions, _ = ensureJobConditionStatus(originalConditions, tc.expectedCondition.Type, tc.expectedCondition.Status, tc.expectedCondition.Reason, "")
 			}
-			if tc.expectedCondition == nil && tc.suspend && len(actual.Status.Conditions) != 0 {
-				t.Errorf("Unexpected conditions %v", actual.Status.Conditions)
+			for _, jc := range originalConditions {
+				if !getCondition(actual, jc.Type, jc.Status, jc.Reason) {
+					copyIrrelevantJobConditionFields(actual.Status.Conditions, originalConditions)
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:748]: I would sort the conditions by name and then do a direct comparison of `actual.Status.Condition` and `original`. Then call `cmp.Diff` with `cmpopts.IgnoreFields`

File: pkg/controller/job/job_controller.go (hunk 5 above)
Comment (adtac) [reply pkg/controller/job/job_controller.go:600]: s/This should never happen/This should not be the case/ :)

Yes, this will happen in the first sync.

File: pkg/controller/job/job_controller.go (hunk 3 above)
Comment (adtac) [reply pkg/controller/job/job_controller.go:591]: this is JobComplete, not JobSuspended

**Comment by adtac** [comment]
is someone else able to take this over? this is an important change but I don't have the time to see it through
